operators/src/upgrade/rest/service.rs

```diff
@@ -125,6 +138,234 @@ pub async fn apply_upgrade() -> Result<HttpResponse, RestError> {
+            .as_str();
+
+        tracing::info!(
+            pod.name = %pod.metadata.name.clone().unwrap_or_default(),
+            node.name = %node_name,
+            "Upgrade starting for data-plane pod"
+        );
+
+        let is_node_cordoned = is_node_cordoned(node_name).await?;
+
+        // Issue node drain command
+        issue_node_drain(node_name).await?;
+
+        // Wait for node drain to complete across the cluster.
+        wait_node_drain().await?;
+
+        // Wait for any rebuild to complete.
+        wait_for_rebuild().await?;
+
+        // restart the data plane pod
+        restart_data_plane(&node_name, pod).await?;
+
+        // Uncordon the drained node
+        if !is_node_cordoned {
+            uncordon_node(node_name).await?;
+        }
+
+        // validate the new pod is up and running
+        verify_data_plane_pod_is_running(node_name).await?;
+
+        // Validate the control plane pod is up and running
+        is_control_plane_running().await?
+    }
+    Ok(())
+}
+
+pub async fn uncordon_node(node_name: &str) -> Result<(), Error> {
+    match UpgradeConfig::get_config()
+        .rest_client()
+        .nodes_api()
+        .delete_node_cordon(node_name, DRAIN_FOR_UPGRADE)
+        .await
+    {
+        Ok(_) => {
+            tracing::info!(node.name = node_name, "Node is uncordoned");
+        }
+        Err(_) => {
+            Error::NodeUncordonError {
+                node_name: node_name.to_string(),
+            };
+        }
+    }
+    Ok(())
+}
+
+/// Issue delete command on dataplane pods.
+pub async fn restart_data_plane(node_name: &str, pod: &Pod) -> Result<(), Error> {
+    let pods: Api<Pod> = Api::namespaced(
+        UpgradeConfig::get_config().k8s_client().client(),
+        UpgradeConfig::get_config().namespace(),
+    );
+    // Deleting the io-engine pod
+    let pod_name = pod.metadata.name.as_ref().unwrap();
+    tracing::info!(
+        pod.name = pod_name,
+        node.name = node_name,
+        "Deleting the pod"
+    );
+    pods.delete(pod_name, &DeleteParams::default())
+        .await?
+        .map_right(|s| tracing::info!(pod.name = pod_name, "Deleted Pod: {:?}", s));
+    Ok(())
+}
+
+/// Wait for the data plane pod to come up on the given node.
+pub async fn wait_node_drain() -> Result<(), Error> {
+    while is_draining().await? {
+        tokio::time::sleep(Duration::from_secs(10_u64)).await;
+    }
+    Ok(())
+}
+
+/// Wait for all the node drain process to complete.
+pub async fn verify_data_plane_pod_is_running(node_name: &str) -> Result<(), Error> {
+    // Validate the new pod is up and running
+    while is_data_plane_pod_running(node_name).await? {
+        tokio::time::sleep(Duration::from_secs(10_u64)).await;
+    }
+    Ok(())
+}
+
+///  Wait for the rebuild to complete if any
+pub async fn wait_for_rebuild() -> Result<(), Error> {
+    // Wait for 30 seconds for any rebuilds to kick in.
+    tokio::time::sleep(Duration::from_secs(30_u64)).await;
```

Comment:
30s might not be enough here, maybe should be conservative and set to 1m.
Later we can add api to bypass cache.